osfoffline/polling_osf_manager/polling.py

```diff
@@ -29,42 +31,84 @@ def __init__(self, user, loop):
+        try:
+            raise future.exception()
+        except (aiohttp.ClientError, asyncio.TimeoutError):
+            logger.error('Unable to connect to the internet')
+            AlertHandler.warn('Unable to connection to OSF, we\'ll try again later.')
+        except asyncio.CancelledError:
+            # Cancellations just mean this thread is exiting
+            return
+
+        # Make sure all our jobs are cancelled
+        if future == self.poll_job:
+            self.process_job.cancel()
+        else:
+            self.poll_job.cancel()
+
+        logger.info('Restarting polling in 5 seconds')
+        # Finally restart our watcher thread in 5 seconds
+        self._loop.call_later(5, self.start)
 
     def start(self):
-        # annoying and weird way to get the remote user from the coroutine
+        remote_user = self._loop.run_until_complete(self.get_remote_user())
 
-        future = asyncio.Future(loop=self._loop)
-        asyncio.async(self.get_remote_user(future), loop=self._loop)
-        self._loop.run_until_complete(future)
-        remote_user = future.result()
+        self.queue = asyncio.Queue(maxsize=15)
 
-        self._loop.call_soon(
-            asyncio.async,
-            self.check_osf(remote_user)
-        )
+        self.process_job = asyncio.ensure_future(self.process_queue())
+        self.poll_job = asyncio.ensure_future(self.check_osf(remote_user))
+
+        self.poll_job.add_done_callback(self.handle_exception)
+        self.process_job.add_done_callback(self.handle_exception)
+
+        return True
 
-    # Only once you have a remote user do we want to check the osf.
-    # thus this coroutine repeatedly tries to get the remote user.
-    # It blocks check_osf from running.
     @asyncio.coroutine
-    def get_remote_user(self, future):
+    def process_queue(self):
+        while True:
+            job = yield from self.queue.get()
+            logger.info('Running {}'.format(job))
+            yield from job.run()
+            self.queue.task_done()
 
+    @asyncio.coroutine
+    def get_remote_user(self):
         url = api_url_for(USERS, user_id=self.user.osf_id)
-        logging.debug(url)
-        while self._keep_running:
+        logger.debug(url)
+        sleep_time = 0
+        while True:
             try:
-                resp = yield from self.osf_query.make_request(url, get_json=True)
-                future.set_result(resp['data'])
-                break
-            except (concurrent.futures._base.TimeoutError, aiohttp.errors.ClientTimeoutError):
-                AlertHandler.warn("Bad Internet Connection")
+                return (yield from self.osf_query.make_request(url, get_json=True))['data']
+            except aiohttp.ClientError:
+                pass
+            sleep_time += 5
+            logger.error('Unable to fetch user from OSF, sleeping for {} seconds'.format(sleep_time))
+            yield from asyncio.sleep(sleep_time)
```

Comment:
Broadly I agree. Two issues brought up during QA would help:
1. A manual "sync now" button that restarts polling on the spot, and,
2. Instead of just saying "up to date", the systray icon menu line that indicates status should say the last updated time (so people know when the last sync actually happened)